### ISSUE TITLE
BE: Improve News & Benefit promo blocks workflow and help text

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#13863]
### User story

**AS A** Homepage Manager role in Drupal
**I WANT** clear workflow and help text when editing custom blocks (News promo, Benefit promo)
**SO THAT** I ensure the homepage works well for Veterans. 

### Notes / background

[Figma for News promo block UI](https://www.figma.com/file/sOMpovTM82TgC08ugF0FxF/News-Promo-Block?type=design&node-id=0-1&mode=design&t=4HX5hbqW1RqCYoA5-0)
- updates are pointed out in comments
- comments will also indicate which changes also apply to Benefit promo block UI

Dave C edited the News promo block and made the following observations:
- We need to do some QA on the help text associated with any of these fields.
- There's no way to "cancel" my edits when I've started editing.
- Revision logs are not required
- It's not clear why the Link Text CTA is a dropdown if the only option is "Read the full article". (Reason: #11172 will add further options.)

**Where to find in CMS:**
Filter block type on [Staging blocks list](https://staging.cms.va.gov/admin/structure/block/block-content)

[Overview of Homepage Editor UX](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/11802) of steps and UI (possibly slightly outdated)

### New AC from Tues, 8/15
**AS A** Homepage Manager (HPM)
**I WANT** to navigate to and edit News and Benefit promo blocks
**SO THAT** I can do my job.
- [x] HPM can
  - [x] Navigate to News and Benefits promo blocks using menus (Structure)

### Existing ACs from planning:
- [x] HPM only has two actions available on News and Benefit promo blocks: "Save draft" and "Publish" (both are available in the `Change to:` dropdown menu)
- [x] News Promo help text
  - [x] Block description: `This is used for identifying the block within the CMS. The description will not be shown on VA.gov.`
  - [x] Link-related text:  Link Text field on the News Promo. Text: `This text tells Veterans what to do next.`
  - [x] "Promo headline" field: make help text, `Use this headline to make it clear to Veterans what the topic of this promo is and why it's relevant to them.`
- [x] Benefit Promo help text
  - [x] Block description: `This is used for identifying the block within the CMS. The description will not be shown on VA.gov.`
  - [x] "Promo headline" field: make help text, `Use this headline to make it clear to Veterans what the topic of this promo is and why it's relevant to them.`

### ACs that need to move to new ticket due to unforeseen complexity re workflows: #14800
- [ ] Revision log required to save News promo block
- [ ] Revision log required to save Benefit promo block

**Comment by thejordanwood** [comment]
@wesrowe I added some comments in the [news promo block Figma file](https://www.figma.com/file/sOMpovTM82TgC08ugF0FxF/News-Promo-Block?type=design&node-id=0%3A1&mode=design&t=fqOkdRBl50BIRhv6-1) to show what we can also carry over to the benefit promo block.

**Comment by wesrowe** [comment]
@dsasser, Jordan and I are trying to figure out help text for the News Promo and Benefit Promo blocks. I thought there was a limitation that we can't change help text for Link-related fields (Link and Link Text) because they come from a module. But there are significant differences between the two already, before @thejordanwood improves them.

[News Promo UI](https://staging.cms.va.gov/block/209?destination=/admin/structure/block/block-content):
<img width="542" alt="image" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/4054752/60d3e030-cacd-4c42-bf53-47a6616a4583">

[Benefit Promo UI](https://staging.cms.va.gov/block/208?destination=/admin/structure/block/block-content):
<img width="520" alt="image" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/4054752/fd747558-4f56-43fd-832b-ff4af8b0325a">

Can you help clarify whether we have free rein to change help text on these 4 (total) help texts?

**Comment by dsasser** [comment]
@wesrowe unfortunately the only help text you have free range over is the News Promo's "Link Text" field. All of the other fields have help text defined by either contrib (linkit: Benefit Promo) or core (link: News Promo).

the News Promo block uses unique "link" and "link text" fields, so managing help text individually is possible out-of-the-box there.

The Benefit Promo block has a different architecture: it instead uses a "CTA" paragraph field. That paragraph is used throughout the site, so we can't change any field properties without it affecting each instance of the paragraph.

**Comment by wesrowe** [comment]
@dsasser, the message that we can't customize the Benefit Promos fields for links is crystal clear, thank you. 

For News Promo, I think I'm seeing conflicting messaging in your comment above(?). Your first sentence says 'the only help text you have free range over is the News Promo's "Link Text" field.' That's clear, but then your second sentence seems to say both News Promo fields can be customized? It's likely you were providing context for the first statement and I'm just not savvy enough to get the nuance... But wanted to confirm.

**Comment by thejordanwood** [comment]
@dsasser @wesrowe Following up on this to get clarity on what help text can be changed. Do I have this right?

**News Promo**
Link (can't be changed)
Link text (CAN change)

**Benefit Promo**
Link (can't be changed)
Link text (can't be changed)

**Comment by wesrowe** [comment]
Jordan and I finalized ACs on 7/10, this is ready for development.

**Comment by jilladams** [comment]
Sprint 89 planning: 
* Blocks use "Delete" instead of "Archive". True in old homepage blocks as well. 
* That pattern doesn't apply in other places in the CMS.
* We can hide the Delete button (code is in place to do this on nodes) / show the Archive button, technically speaking.
* We expect that Editors will use the same News promo block and edit it over time, rather than creating new blocks. 
    * We expect that Editors will not mess with the EntityQueue. 
* Governance-wise: 
    * Scope change: we don't want editors to have the ability to take down the News promo block. Draft / Publish should be the only available workflow options.
    * What makes most sense to Editors? Josh Tuscher question

**Comment by wesrowe** [comment]
Wes to regroup, figure out if we want to mess with "cancel" paradigm, update for help text we can change.

**Comment by chri5tia** [comment]
TL;DR: Delete button shouldn't be showing up for anyone except us, so worrying about that button is moot.

Observation:

Currently no users are able to delete block content except admins. I checked the permissions. Delete shouldn't show up for anyone except the `administrator` role.

Screenshot shows delete block permissions for `content_admins`. I
<img width="1176" alt="Screenshot 2023-08-08 at 11 27 50 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/22764938/bb7150bb-0f51-4d0b-a4e7-da42f0f7841f">
 checked the other roles as well.

**Comment by chri5tia** [comment]
@jilladams @wesrowe Which roles should see "Publish"? My understanding from the latest mockup comments and this ticket that the two buttons should say "Save draft" and "Publish". Which roles for each of these options?

**Comment by chri5tia** [comment]
@wesrowe @jilladams AC's mention that the buttons are "Save draft" and "Publish". Does this mean that the state change drop down menu options are draft and publish? This is dependent on the user's role based on the workflow already defined and applied to both nodes and blocks. As for buttons at the bottom, there are "Save" or "Save as draft and continue". I am not sure what the third AC is asking for.

<img width="446" alt="Screenshot 2023-08-09 at 12 49 19 AM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/22764938/b2749402-cd4f-487c-8bfa-5702181518a3">

**Comment by chri5tia** [comment]
## Engineering notes for making revision log required
https://www.drupal.org/project/block_form_alter provides hooks for blocks

**Comment by wesrowe** [comment]
@chri5tia, it seems to me that the dropdown having "draft" and "published" would meet the AC. The key is that they can save a draft, and they can NOT delete or unpublish the block.

**Comment by chri5tia** [comment]
This is ready for a preliminary review sans the log message required, which I am still strategizing about. This may include submitting a contrib module and I'd like more time to figure it out. I propose we move making the log message required on blocks as a separate stretch ticket that I can start working on. 

If we do that then this is ready to review.

**Comment by chri5tia** [comment]
Targeting a block will require a slightly [different method than what was used for nodes and taxonomy](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/749ae78cb1df492da5b449a1f3fce2f56d924e86/docroot/modules/custom/va_gov_workflow/src/EventSubscriber/EntityEventSubscriber.php#L241-L260) but it shouldn't require a whole module. I had found one that implemented new alter hooks that would make it smooth but not necessarily perfect for our product.

**Comment by jilladams** [comment]
(I've given Dave Conlon's user the Homepage Manager role in prod, so tomorrow we'll be able to use that user to test in Staging.)

**Comment by wesrowe** [comment]
@jilladams, update on the Homepage Manager role – Christia and I added ACs to the current-sprint ticket for promo blocks to make sure that HPMs can find the blocks in order to edit them. Permissions seem to have been a miss on the original work, we created a test user and they couldn't do anything with promo blocks. But Christia said it's a simple add. Code probably won't merge till tomorrow though.

Rev log ACs moved to a new ticket (#14800), we can refine tomorrow if there's time or just wait till BE refinement.

**Comment by jilladams** [comment]
Worth double checking against #11160 (ticket) and PR https://github.com/department-of-veterans-affairs/va.gov-cms/pull/11534 for what we defined there / implemented, vs. what's missing now. (Daniel did that original work.)

**Comment by chri5tia** [comment]
Changes to previous homepage permissions include disallowing archiving the block, and allowing access to structure in the menu to get to entity queues. I am testing now whether or not the latter was needed.

**Comment by wesrowe** [comment]
@chri5tia, I've lost track of this ticket, but I see that a couple days ago it went back to In Progress. What's the status?

**Comment by chri5tia** [comment]
I'm working on it today/tonight.

**Comment by chri5tia** [comment]
The workflow is already configured in the UI to include these two specific promo blocks (and a couple others that we also don't want non-admins to archive). This then uses the permissions by role to remove the archive option. This method to accomplish this ticket last sprint was rejected.

I can't figure out how to add this seemingly existent functionality into the custom code that was explained in [this slack thread](https://dsva.slack.com/archives/C52CL1PKQ/p1692214385800349?thread_ts=1692138925.970619&cid=C52CL1PKQ) and how it isn't redundant. 

I've tried a few things but wasn't able to get it to work. Will probably pair up with someone.

**Comment by chri5tia** [comment]
This part of the entity event subscriber for the workflow module needs to be adapted to include the promo custom blocks:

```
  private function removeArchiveOption(FormBaseAlterEvent $event) {
    $form = &$event->getForm();
    $form_state = $event->getFormState();
    if ($form_state->getFormObject() instanceof NodeForm) {
      /** @var \Drupal\node\NodeInterface $node **/
      $node = $form_state->getFormObject()->getEntity();
      $bundle = $node->bundle();
      $is_admin = $this->userPermsService->hasAdminRole();
      $is_archiveable = $this->workflowContentControl->isBundleArchiveableByNonAdmins($bundle);
      if (!$is_admin && !$is_archiveable) {
        unset($form['moderation_state']['widget'][0]['state']['#options']['archived']);
      }
    }
  }
```

I can't figure out the machine name but they are block/208 and block/209 custom_block_news_promo I think.

See this slack conversation:
https://dsva.slack.com/archives/C52CL1PKQ/p1692214324958539?thread_ts=1692138925.970619&cid=C52CL1PKQ

**Comment by chri5tia** [comment]
@jilladams This PR will be ready to review shortly and should close by EOD

**Comment by chri5tia** [comment]
Not sure at the moment how this got reopened, but I am closing it again. Work is complete.
